pkg/controller/pxcrestore/manager.go

```diff
@@ -0,0 +1,259 @@
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	api "github.com/percona/percona-xtradb-cluster-operator/pkg/apis/pxc/v1"
+	"github.com/percona/percona-xtradb-cluster-operator/pkg/k8s"
+	"github.com/percona/percona-xtradb-cluster-operator/pkg/pxc/backup"
+	"github.com/percona/percona-xtradb-cluster-operator/pkg/pxc/backup/storage"
+)
+
+// TODO: choose a better name for this interface
+type RestoreManager interface {
+	Init(ctx context.Context) error
+	Job() (*batchv1.Job, error)
+	PITRJob() (*batchv1.Job, error)
+	Finalize(ctx context.Context) error
+	Validate(ctx context.Context) error
+}
+
+type s3 struct{ *restoreManagerOptions }
+
+func (s *s3) Init(context.Context) error     { return nil }
+func (s *s3) Finalize(context.Context) error { return nil }
+
+func (s *s3) Job() (*batchv1.Job, error) {
+	return backup.S3RestoreJob(s.cr, s.bcp, strings.TrimPrefix(s.bcp.Status.Destination, api.AwsBlobStoragePrefix), s.cluster, false)
+}
+
+func (s *s3) PITRJob() (*batchv1.Job, error) {
+	return backup.S3RestoreJob(s.cr, s.bcp, strings.TrimPrefix(s.bcp.Status.Destination, api.AwsBlobStoragePrefix), s.cluster, true)
+}
+
+func (s *s3) Validate(ctx context.Context) error {
+	sec := corev1.Secret{}
+	err := s.k8sClient.Get(ctx,
+		types.NamespacedName{Name: s.bcp.Status.S3.CredentialsSecret, Namespace: s.bcp.Namespace}, &sec)
+	if err != nil {
+		return errors.Wrap(err, "failed to get secret")
+	}
+
+	accessKeyID := string(sec.Data["AWS_ACCESS_KEY_ID"])
+	secretAccessKey := string(sec.Data["AWS_SECRET_ACCESS_KEY"])
+	ep := s.bcp.Status.S3.EndpointURL
+	bucket, prefix := s.bcp.Status.S3.BucketAndPrefix()
+	verifyTLS := true
+	if s.bcp.Status.VerifyTLS != nil && !*s.bcp.Status.VerifyTLS {
+		verifyTLS = false
+	}
+	if s.cluster.Spec.Backup != nil && len(s.cluster.Spec.Backup.Storages) > 0 {
+		storage, ok := s.cluster.Spec.Backup.Storages[s.bcp.Spec.StorageName]
+		if ok && storage.VerifyTLS != nil {
+			verifyTLS = *storage.VerifyTLS
+		}
+	}
+	s3cli, err := s.newStorageClient(&storage.S3Options{
+		Endpoint:        ep,
+		AccessKeyID:     accessKeyID,
+		SecretAccessKey: secretAccessKey,
+		BucketName:      bucket,
+		Prefix:          prefix,
+		Region:          s.bcp.Status.S3.Region,
+		VerifyTLS:       verifyTLS,
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to create s3 client")
+	}
+	dest := s.bcp.Status.Destination
+	dest = strings.TrimPrefix(dest, api.AwsBlobStoragePrefix)
+	dest = strings.TrimPrefix(dest, bucket+"/")
+	if prefix != "" {
+		dest = strings.TrimPrefix(dest, prefix)
+		dest = strings.TrimPrefix(dest, "/")
+	}
+	dest = strings.TrimSuffix(dest, "/") + "/"
+
+	objs, err := s3cli.ListObjects(ctx, dest)
+	if err != nil {
+		return errors.Wrap(err, "failed to list objects")
+	}
+	if len(objs) == 0 {
+		return errors.New("backup not found")
+	}
+
+	return nil
+}
+
+type pvc struct{ *restoreManagerOptions }
+
+func (s *pvc) Validate(context.Context) error { return nil }
```

Comment:
what if PVC is already deleted? Shouldn't we check that before terminating the cluster?

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/1403/commits/f45ed5444c85476ef2f78369b576059ede9acbae